diff -urN --binary -x builds -x buildd libarchive-3.0.2.orig/libarchive/archive_write_disk_windows.c libarchive-3.0.2/libarchive/archive_write_disk_windows.c
--- libarchive-3.0.2.orig/libarchive/archive_write_disk_windows.c	2011-12-24 18:56:28.000000000 +0100
+++ libarchive-3.0.2/libarchive/archive_write_disk_windows.c	2011-12-28 12:01:13.595083500 +0100
@@ -542,11 +542,24 @@
 {
 	static BOOLEAN (WINAPI *f)(LPWSTR, LPWSTR, LPSECURITY_ATTRIBUTES);
 	static int set;
+	BOOL ret;
 	if (!set) {
 		set = 1;
 		f = la_GetFunctionKernel32("CreateHardLinkW");
 	}
-	return f == NULL ? 0 : (*f)(linkname, target, NULL);
+	if (!f)
+		return 0;
+	ret = (*f)(linkname, target, NULL);
+	if (!ret)
+	{
+		/* Under windows 2000, it is necessary to remove the "\\?\" prefix */
+		if (!wcsncmp(linkname,L"\\\\?\\",4))
+			linkname += 4;
+		if (!wcsncmp(target,L"\\\\?\\",4))
+			target += 4;
+		ret = (*f)(linkname, target, NULL);
+	}
+	return ret;
 }
 
 static int
