Index: CMakeLists.txt
===================================================================
--- CMakeLists.txt	(revision 2950)
+++ CMakeLists.txt	(working copy)
@@ -487,8 +487,19 @@
 CHECK_FUNCTION_EXISTS_GLIBC(fdopendir HAVE_FDOPENDIR)
 CHECK_FUNCTION_EXISTS_GLIBC(fork HAVE_FORK)
 CHECK_FUNCTION_EXISTS_GLIBC(fstat HAVE_FSTAT)
-CHECK_FUNCTION_EXISTS_GLIBC(fstatat HAVE_FSTATAT)
-CHECK_FUNCTION_EXISTS_GLIBC(fstatfs HAVE_FSTATFS)
+
+# Need to turn implicit declaration warnings into errors because
+# that's what the build will do. For LSB compilers, a deprecation
+# warning is issued for statfs() and fstatfs(), while fstatat()
+# is missing from some LSB headers by mistake.
+set(CMAKE_REQUIRED_FLAGS "-Wimplicit -Werror")
+CHECK_C_SOURCE_COMPILES(
+  "#include <sys/stat.h>\n#include <fcntl.h>\nint main() { struct stat buf; return fstatat(0,\"\",&buf,0); }"
+  HAVE_FSTATAT)
+CHECK_C_SOURCE_COMPILES(
+  "#include <sys/stat.h>\n#include <unistd.h>\nint main() { struct stat buf; return fstatfs(0,&buf); }"
+  HAVE_FSTATFS)
+
 CHECK_FUNCTION_EXISTS_GLIBC(fstatvfs HAVE_FSTATVFS)
 CHECK_FUNCTION_EXISTS_GLIBC(ftruncate HAVE_FTRUNCATE)
 CHECK_FUNCTION_EXISTS_GLIBC(futimens HAVE_FUTIMENS)
@@ -525,7 +536,10 @@
 CHECK_FUNCTION_EXISTS_GLIBC(setenv HAVE_SETENV)
 CHECK_FUNCTION_EXISTS_GLIBC(setlocale HAVE_SETLOCALE)
 CHECK_FUNCTION_EXISTS_GLIBC(sigaction HAVE_SIGACTION)
-CHECK_FUNCTION_EXISTS_GLIBC(statfs HAVE_STATFS)
+# Need to build for this test to catch LSB deprecation warnings - see above
+CHECK_C_SOURCE_COMPILES(
+  "#include <sys/stat.h>\n#include <unistd.h>\nint main() { struct stat buf; return statfs(\"\",&buf); }"
+  HAVE_STATFS)
 CHECK_FUNCTION_EXISTS_GLIBC(statvfs HAVE_STATVFS)
 CHECK_FUNCTION_EXISTS_GLIBC(strchr HAVE_STRCHR)
 CHECK_FUNCTION_EXISTS_GLIBC(strdup HAVE_STRDUP)
