Index: libarchive/archive_read_open_filename.c
===================================================================
--- libarchive/archive_read_open_filename.c	(リビジョン 656)
+++ libarchive/archive_read_open_filename.c	(作業コピー)
@@ -248,7 +248,8 @@
 	 * Here, I flush everything except for regular files and
 	 * device nodes.
 	 */
-	if (!S_ISREG(mine->st_mode)
+	if (mine->fd != -1
+	    && !S_ISREG(mine->st_mode)
 	    && !S_ISCHR(mine->st_mode)
 	    && !S_ISBLK(mine->st_mode)) {
 		ssize_t bytesRead;
@@ -258,7 +259,7 @@
 		} while (bytesRead > 0);
 	}
 	/* If a named file was opened, then it needs to be closed. */
-	if (mine->filename[0] != '\0')
+	if (mine->fd != -1 && mine->filename[0] != '\0')
 		close(mine->fd);
 	if (mine->buffer != NULL)
 		free(mine->buffer);
