Index: libarchive/test/test_read_format_7zip.c
===================================================================
--- libarchive/test/test_read_format_7zip.c	(revision 4213)
+++ libarchive/test/test_read_format_7zip.c	(working copy)
@@ -26,8 +26,8 @@
 __FBSDID("$FreeBSD");
 
 /*
- * Extract a non-encorded file.
- * The header of the 7z archive files is not encdoed.
+ * Extract a non-encoded file.
+ * The header of the 7z archive files is not encoded.
  */
 static void
 test_copy()
@@ -139,7 +139,7 @@
 
 /*
  * Extract an encoded file.
- * The header of the 7z archive files is not encdoed.
+ * The header of the 7z archive files is not encoded.
  */
 static void
 test_plain_header(const char *refname)
@@ -180,7 +180,7 @@
 
 /*
  * Extract multi files.
- * The header of the 7z archive files is encdoed with LZMA.
+ * The header of the 7z archive files is encoded with LZMA.
  */
 static void
 test_extract_all_files(const char *refname)
@@ -255,7 +255,7 @@
 
 /*
  * Extract last file.
- * The header of the 7z archive files is encdoed with LZMA.
+ * The header of the 7z archive files is encoded with LZMA.
  */
 static void
 test_extract_last_file(const char *refname)
@@ -323,7 +323,7 @@
 }
 
 /*
- * Extract a mixed archive file which has both  LZMA and LZMA2 encoded files.
+ * Extract a mixed archive file which has both LZMA and LZMA2 encoded files.
  *  LZMA: file1, file2, file3, file4
  *  LZMA2: zfile1, zfile2, zfile3, zfile4
  */
Index: libarchive/test/test_read_format_tar_filename.c
===================================================================
--- libarchive/test/test_read_format_tar_filename.c	(revision 4213)
+++ libarchive/test/test_read_format_tar_filename.c	(working copy)
@@ -34,7 +34,7 @@
  * - the filename of second file is stored in UTF-8.
  *
  * Whenever hdrcharset option is specified, we will correctly read the
- * filename of sencod file, which is stored in UTF-8 by default.
+ * filename of second file, which is stored in UTF-8 by default.
  */
 
 static void
