Index: libarchive/archive_write.c
===================================================================
--- libarchive/archive_write.c	(revision 3741)
+++ libarchive/archive_write.c	(working copy)
@@ -619,9 +619,7 @@
 	if (ret < ARCHIVE_OK && ret != ARCHIVE_WARN)
 		return (ret);
 
-	if (a->skip_file_dev != 0 &&
-	    archive_entry_dev(entry) == a->skip_file_dev &&
-	    a->skip_file_ino != 0 &&
+	if (archive_entry_dev(entry) == a->skip_file_dev &&
 	    archive_entry_ino64(entry) == a->skip_file_ino) {
 		archive_set_error(&a->archive, 0,
 		    "Can't add archive to itself");
