diff --git a/CMakeLists.txt b/CMakeLists.txt
index 11b7139..6a5f792 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -15,7 +15,7 @@ endif()
 #   RelWithDebInfo : Release build with Debug Info
 #   MinSizeRel     : Release Min Size build
 IF(NOT CMAKE_BUILD_TYPE)
-  SET(CMAKE_BUILD_TYPE "Release" CACHE STRING "Build Type" FORCE)
+  SET(CMAKE_BUILD_TYPE "Debug" CACHE STRING "Build Type" FORCE)
 ENDIF(NOT CMAKE_BUILD_TYPE)
 # Set a value type to properly display CMAKE_BUILD_TYPE on GUI if the
 # value type is "UNINITIALIZED".
diff --git a/libarchive/archive_string.c b/libarchive/archive_string.c
index 87f9288..f71692d 100644
--- a/libarchive/archive_string.c
+++ b/libarchive/archive_string.c
@@ -1997,7 +1997,7 @@ static int
 iconv_strncat_in_locale(struct archive_string *as, const void *_p,
     size_t length, struct archive_string_conv *sc)
 {
-	ICONV_CONST char *itp;
+	char *itp;
 	size_t remaining;
 	iconv_t cd;
 	char *outp;
diff --git a/libarchive/test/test_read_format_rar.c b/libarchive/test/test_read_format_rar.c
index e4e57f6..5870323 100644
--- a/libarchive/test/test_read_format_rar.c
+++ b/libarchive/test/test_read_format_rar.c
@@ -721,8 +721,8 @@ DEFINE_TEST(test_read_format_rar_ppmd_lzss_conversion)
 DEFINE_TEST(test_read_format_rar_binary)
 {
   const char reffile[] = "test_read_format_rar_binary_data.rar";
-  char file1_buff[1048576];
-  int file1_size = sizeof(file1_buff);
+  char *file1_buff = malloc(1048576);
+  int file1_size = 1048576;
   const char file1_test_txt[] = "\x37\xef\xb2\xbe\x33\xf6\xcc\xcb\xee\x2a\x10"
                                 "\x9d\x2e\x01\xe9\xf6\xf9\xe5\xe6\x67\x0c\x2b"
                                 "\xd8\x6b\xa0\x26\x9a\xf7\x93\x87\x42\xf1\x08"
@@ -777,6 +777,8 @@ DEFINE_TEST(test_read_format_rar_binary)
   assertEqualInt(2, archive_file_count(a));
   assertEqualIntA(a, ARCHIVE_OK, archive_read_close(a));
   assertEqualInt(ARCHIVE_OK, archive_read_free(a));
+
+  free(file1_buff);
 }
 
 DEFINE_TEST(test_read_format_rar_windows)
