--- /var/tmp/paludis/app-arch-libarchive-scm/work/tar/test/test_option_gid_gname.c~	2010-07-31 19:45:40.000000000 +0100
+++ /var/tmp/paludis/app-arch-libarchive-scm/work/tar/test/test_option_gid_gname.c	2010-07-31 19:45:42.000000000 +0100
@@ -57,7 +57,7 @@
 	/* Again with just --gid */
 	failure("Error invoking %s c", testprog);
 	assertEqualInt(0,
-	    systemf("%s cf archive3 --gid=17 --format=ustar file >stdout3.txt 2>stderr3.txt",
+	    systemf("%s cf archive3 --gid=17 --gname= --format=ustar file >stdout3.txt 2>stderr3.txt",
 		testprog));
 	assertEmptyFile("stdout3.txt");
 	assertEmptyFile("stderr3.txt");
