Index: libarchive/archive_read_disk_entry_from_file.c
===================================================================
--- libarchive/archive_read_disk_entry_from_file.c	(revision 2526)
+++ libarchive/archive_read_disk_entry_from_file.c	(working copy)
@@ -387,7 +387,7 @@
 		list_size = listxattr(path, NULL, 0);
 
 	if (list_size == -1) {
-		if (errno == ENOTSUP)
+		if (errno == ENOTSUP || errno == EOPSYS)
 			return (ARCHIVE_OK);
 		archive_set_error(&a->archive, errno,
 			"Couldn't list extended attributes");
Index: libarchive/archive_write_disk.c
===================================================================
--- libarchive/archive_write_disk.c	(revision 2526)
+++ libarchive/archive_write_disk.c	(working copy)
@@ -2456,7 +2456,7 @@
 				    name, value, size, 0);
 			}
 			if (e == -1) {
-				if (errno == ENOTSUP) {
+				if (errno == ENOTSUP || errno == EOPSYS) {
 					if (!warning_done) {
 						warning_done = 1;
 						archive_set_error(&a->archive, errno,
@@ -2523,7 +2523,7 @@
 				    namespace, name, value, size);
 			}
 			if (e != (int)size) {
-				if (errno == ENOTSUP) {
+				if (errno == ENOTSUP || errno == EOPSYS) {
 					if (!warning_done) {
 						warning_done = 1;
 						archive_set_error(&a->archive, errno,
