--- CMakeLists.txt	2011-01-04 19:18:09.382620531 +1100
+++ CMakeLists.txt.patched	2011-01-04 19:42:02.738404289 +1100
@@ -549,8 +549,11 @@
 CHECK_SYMBOL_EXISTS(vprintf          "stdio.h"    HAVE_VPRINTF)
 CHECK_SYMBOL_EXISTS(cygwin_conv_path "sys/cygwin.h" HAVE_CYGWIN_CONV_PATH)
 
-CHECK_SYMBOL_EXISTS(major            "sys/mkdev.h"     MAJOR_IN_MKDEV)
-CHECK_SYMBOL_EXISTS(major            "sys/sysmacros.h" MAJOR_IN_SYSMACROS)
+GET_FILENAME_COMPONENT(ccomp_lsbcheck ${CMAKE_C_COMPILER} NAME)
+IF(NOT ccomp_lsbcheck STREQUAL "lsbcc")
+  CHECK_SYMBOL_EXISTS(major            "sys/mkdev.h"     MAJOR_IN_MKDEV)
+  CHECK_SYMBOL_EXISTS(major            "sys/sysmacros.h" MAJOR_IN_SYSMACROS)
+ENDIF()
 
 IF(HAVE_STRERROR_R)
   SET(HAVE_DECL_STRERROR_R 1)
