--- a/libarchive/archive_write_disk.c
+++ b/libarchive/archive_write_disk.c
@@ -434,6 +434,7 @@
 		if (ret != ARCHIVE_OK)
 			goto done;
 	}
+	a->restore_pwd = -1;
 #if defined(HAVE_FCHDIR) && defined(PATH_MAX)
 	/* If path exceeds PATH_MAX, shorten the path. */
 	edit_deep_directories(a);
@@ -873,8 +874,6 @@
 	int ret;
 	char *tail = a->name;
 
-	a->restore_pwd = -1;
-
 	/* If path is short, avoid the open() below. */
 	if (strlen(tail) <= PATH_MAX)
 		return;
