Index: libarchive/test/test_tar_large.c
===================================================================
--- libarchive/test/test_tar_large.c	(revision 842)
+++ libarchive/test/test_tar_large.c	(working copy)
@@ -68,10 +68,10 @@
 };
 
 /* The following size definitions simplify things below. */
-#define KB ((off_t)1024)
-#define MB ((off_t)1024 * KB)
-#define GB ((off_t)1024 * MB)
-#define TB ((off_t)1024 * GB)
+#define KB ((int64_t)1024)
+#define MB ((int64_t)1024 * KB)
+#define GB ((int64_t)1024 * MB)
+#define TB ((int64_t)1024 * GB)
 
 #if ARCHIVE_VERSION_NUMBER < 2000000
 static ssize_t	memory_read_skip(struct archive *, void *, size_t request);
@@ -202,7 +202,7 @@
 DEFINE_TEST(test_tar_large)
 {
 	/* The sizes of the entries we're going to generate. */
-	static off_t tests[] = {
+	static int64_t tests[] = {
 		/* Test for 32-bit signed overflow. */
 		2 * GB - 1, 2 * GB, 2 * GB + 1,
 		/* Test for 32-bit unsigned overflow. */
