--- libarchive/archive_match.c.orig	2015-01-13 12:03:38.000000000 +0100
+++ libarchive/archive_match.c	2015-02-06 14:04:37.000000000 +0100
@@ -585,6 +585,12 @@
 		archive_read_free(ar);
 		return (r);
 	}
+	r = archive_read_support_format_empty(ar);
+	if (r != ARCHIVE_OK) {
+		archive_copy_error(&(a->archive), ar);
+		archive_read_free(ar);
+		return (r);
+	}
 	if (mbs)
 		r = archive_read_open_filename(ar, pathname, 512*20);
 	else
@@ -595,7 +601,7 @@
 		return (r);
 	}
 	r = archive_read_next_header(ar, &ae);
-	if (r != ARCHIVE_OK) {
+	if (r != ARCHIVE_OK && r != ARCHIVE_EOF) {
 		archive_copy_error(&(a->archive), ar);
 		archive_read_free(ar);
 		return (r);
@@ -603,6 +609,7 @@
 
 	archive_string_init(&as);
 
+	if (r != ARCHIVE_EOF)
 	while ((r = archive_read_data_block(ar, &buff, &size, &offset))
 	    == ARCHIVE_OK) {
 		const char *b = (const char *)buff;
