Index: util.c
===================================================================
--- util.c	(revision 1365)
+++ util.c	(working copy)
@@ -286,10 +286,8 @@
 	FILE *f;
 	char *buff, *buff_end, *line_start, *line_end, *p;
 	size_t buff_length, new_buff_length, bytes_read, bytes_wanted;
-	int separator;
 	int ret;
 
-	separator = bsdtar->option_null ? '\0' : '\n';
 	ret = 0;
 
 	if (strcmp(pathname, "-") == 0)
@@ -310,12 +308,19 @@
 		buff_end += bytes_read;
 		/* Process all complete lines in the buffer. */
 		while (line_end < buff_end) {
-			if (*line_end == separator) {
-				*line_end = '\0';
+			if (bsdtar->option_null && *line_end == '\0') {
 				if ((*process)(bsdtar, line_start) != 0)
 					ret = -1;
 				line_start = line_end + 1;
 				line_end = line_start;
+			} else if (*line_end == '\r' || *line_end == '\n') {
+				*line_end = '\0';
+				if (line_end - line_start > 1) {
+					if ((*process)(bsdtar, line_start) != 0)
+						ret = -1;
+				}
+				line_start = line_end + 1;
+				line_end = line_start;
 			} else
 				line_end++;
 		}
