Index: util.c
===================================================================
--- util.c	(revision 644)
+++ util.c	(working copy)
@@ -517,6 +517,27 @@
 	while (name[0] == '/' && name[1] == '/')
 		name++;
 
+	/* Strip leading drive letter from files archived on Windows. */
+	/* TODO: should handle absolute paths on Windows; that requires
+	 * work in libarchive. */
+	if (!bsdtar->option_absolute_paths
+	    && ((name[0] >= 'A' && name[0] <= 'Z')
+		|| (name[0] >= 'a' && name[0] <= 'z'))
+	    && name[1] == ':'
+	    && (name[2] == '/' || name[2] == '\\'))
+	{
+		/* Generate a warning the first time this happens. */
+		if (!bsdtar->warned_lead_slash) {
+			bsdtar_warnc(bsdtar, 0,
+			    "Removing leading drive letter from member names");
+			bsdtar->warned_lead_slash = 1;
+		}
+		name += 3;
+		/* Special case: Stripping leading '/' from "/" yields ".". */
+		if (*name == '\0')
+			name = ".";
+	}
+
 	/* Strip leading '/' unless user has asked us not to. */
 	if (name[0] == '/' && !bsdtar->option_absolute_paths) {
 		/* Generate a warning the first time this happens. */
