Index: tar/read.c
===================================================================
--- tar/read.c	(revision 3759)
+++ tar/read.c	(working copy)
@@ -234,7 +234,7 @@
 		/*
 		 * Exclude entries that are too old.
 		 */
-		if (bsdtar->newer_ctime_sec > 0) {
+		if (bsdtar->newer_ctime_filter) {
 			/* Use ctime if format provides, else mtime. */
 			if (archive_entry_ctime_is_set(entry)) {
 				sec = archive_entry_ctime(entry);
@@ -252,7 +252,7 @@
 			    && nsec <= bsdtar->newer_ctime_nsec)
 				continue; /* Too old, skip it. */
 		}
-		if (bsdtar->newer_mtime_sec > 0) {
+		if (bsdtar->newer_mtime_filter) {
 			if (archive_entry_mtime_is_set(entry)) {
 				sec = archive_entry_mtime(entry);
 				nsec = archive_entry_mtime_nsec(entry);
Index: tar/bsdtar.c
===================================================================
--- tar/bsdtar.c	(revision 3759)
+++ tar/bsdtar.c	(working copy)
@@ -395,6 +395,7 @@
 		 * TODO: Add corresponding "older" options to reverse these.
 		 */
 		case OPTION_NEWER_CTIME: /* GNU tar */
+			bsdtar->newer_ctime_filter = 1;
 			bsdtar->newer_ctime_sec = get_date(now, bsdtar->argument);
 			break;
 		case OPTION_NEWER_CTIME_THAN:
@@ -403,12 +404,14 @@
 				if (stat(bsdtar->argument, &st) != 0)
 					lafe_errc(1, 0,
 					    "Can't open file %s", bsdtar->argument);
+				bsdtar->newer_ctime_filter = 1;
 				bsdtar->newer_ctime_sec = st.st_ctime;
 				bsdtar->newer_ctime_nsec =
 				    ARCHIVE_STAT_CTIME_NANOS(&st);
 			}
 			break;
 		case OPTION_NEWER_MTIME: /* GNU tar */
+			bsdtar->newer_mtime_filter = 1;
 			bsdtar->newer_mtime_sec = get_date(now, bsdtar->argument);
 			break;
 		case OPTION_NEWER_MTIME_THAN:
@@ -417,6 +420,7 @@
 				if (stat(bsdtar->argument, &st) != 0)
 					lafe_errc(1, 0,
 					    "Can't open file %s", bsdtar->argument);
+				bsdtar->newer_mtime_filter = 1;
 				bsdtar->newer_mtime_sec = st.st_mtime;
 				bsdtar->newer_mtime_nsec =
 				    ARCHIVE_STAT_MTIME_NANOS(&st);
Index: tar/bsdtar.h
===================================================================
--- tar/bsdtar.h	(revision 3759)
+++ tar/bsdtar.h	(working copy)
@@ -46,8 +46,10 @@
 	const char	 *create_format; /* -F format */
 	char		 *pending_chdir; /* -C dir */
 	const char	 *names_from_file; /* -T file */
+	int		  newer_ctime_filter; /* --newer/--newer-than */
 	time_t		  newer_ctime_sec; /* --newer/--newer-than */
 	long		  newer_ctime_nsec; /* --newer/--newer-than */
+	int		  newer_mtime_filter; /* --newer-mtime/--newer-mtime-than */
 	time_t		  newer_mtime_sec; /* --newer-mtime */
 	long		  newer_mtime_nsec; /* --newer-mtime-than */
 	int		  bytes_per_block; /* -b block_size */
Index: tar/write.c
===================================================================
--- tar/write.c	(revision 3759)
+++ tar/write.c	(working copy)
@@ -1105,7 +1105,7 @@
 	/*
 	 * If this file/dir is excluded by a time comparison, skip it.
 	 */
-	if (bsdtar->newer_ctime_sec > 0) {
+	if (bsdtar->newer_ctime_filter) {
 		if (st->st_ctime < bsdtar->newer_ctime_sec)
 			return (0); /* Too old, skip it. */
 		if (st->st_ctime == bsdtar->newer_ctime_sec
@@ -1113,7 +1113,7 @@
 		    <= bsdtar->newer_ctime_nsec)
 			return (0); /* Too old, skip it. */
 	}
-	if (bsdtar->newer_mtime_sec > 0) {
+	if (bsdtar->newer_mtime_filter) {
 		if (st->st_mtime < bsdtar->newer_mtime_sec)
 			return (0); /* Too old, skip it. */
 		if (st->st_mtime == bsdtar->newer_mtime_sec
