Index: libarchive/archive_acl.c
===================================================================
--- libarchive/archive_acl.c    (revision 4196)
+++ libarchive/archive_acl.c    (working copy)
@@ -531,6 +531,7 @@
                ap = acl->acl_head;
                count = 0;
                while (ap != NULL) {
+                       r = 0;
                        if ((ap->type & ARCHIVE_ENTRY_ACL_TYPE_DEFAULT) != 0 &&
                            (r = archive_mstring_get_wcs(a, &ap->name,
                                        &wname)) == 0) {
Index: libarchive/archive_string_sprintf.c
===================================================================
--- libarchive/archive_string_sprintf.c (revision 4196)
+++ libarchive/archive_string_sprintf.c (working copy)
@@ -40,6 +40,7 @@
  */
 
 #include <stdio.h>
+#include <errno.h>
 
 #include "archive_string.h"
 #include "archive_private.h"
