Index: libarchive/archive_entry_stat.c
===================================================================
--- libarchive/archive_entry_stat.c	(revision 3759)
+++ libarchive/archive_entry_stat.c	(working copy)
@@ -41,7 +41,7 @@
 {
 	struct stat *st;
 	if (entry->stat == NULL) {
-		entry->stat = malloc(sizeof(*st));
+		entry->stat = calloc(1, sizeof(*st));
 		if (entry->stat == NULL)
 			return (NULL);
 		entry->stat_valid = 0;
