diff -urN --binary -x builds -x buildd libarchive-3.0.2.orig/libarchive/archive_windows.c libarchive-3.0.2/libarchive/archive_windows.c
--- libarchive-3.0.2.orig/libarchive/archive_windows.c	2011-12-24 18:56:28.000000000 +0100
+++ libarchive-3.0.2/libarchive/archive_windows.c	2011-12-26 14:01:28.352243500 +0100
@@ -135,10 +135,10 @@
 	l = GetFullPathNameW(wname, 0, NULL, NULL);
 	if (l == 0)
 		return (NULL);
-	wnp = malloc(l * sizeof(wchar_t));
+	wnp = malloc((l+3) * sizeof(wchar_t));
 	if (wnp == NULL)
 		return (NULL);
-	len = GetFullPathNameW(wname, l, wnp, NULL);
+	len = GetFullPathNameW(wname, l+3, wnp, NULL);
 	wn = wnp;
 
 	if (wnp[0] == L'\\' && wnp[1] == L'\\' &&
