Index: libarchive/archive_read_disk_entry_from_file.c
===================================================================
--- libarchive/archive_read_disk_entry_from_file.c	(revision 2024)
+++ libarchive/archive_read_disk_entry_from_file.c	(working copy)
@@ -243,6 +243,11 @@
 #if HAVE_ACL_GET_LINK_NP
 	else if (!a->follow_symlinks)
 		acl = acl_get_link_np(accpath, ACL_TYPE_ACCESS);
+#else
+	else if ((!a->follow_symlinks) && (S_ISLNK(archive_entry_mode(entry))))
+		/* We can't get the ACL of a symlink, so we assume it can't
+		   have one. */
+		acl = NULL;
 #endif
 	else
 		acl = acl_get_file(accpath, ACL_TYPE_ACCESS);
