Index: libarchive/archive_write_disk.c
===================================================================
--- libarchive/archive_write_disk.c	(revision 659)
+++ libarchive/archive_write_disk.c	(working copy)
@@ -1593,6 +1593,22 @@
 			*dest++ = '/';
 		while (*src != '\0' && *src != '/') {
 			*dest++ = *src++;
+#if defined(__WIN32__) || defined(_WIN32) || defined(__WIN32)
+			// If compiled for a MS-Windows platform,
+			// rewrite characters that are unusable.
+			//
+			// See also:
+			// http://msdn.microsoft.com/library/?url=/library/en-us/fileio/fs/naming_a_file.asp
+			if (
+				(*dest == ':') ||
+				(*dest == '*') ||
+				(*dest == '?') ||
+				(*dest == '"') ||
+				(*dest == '<') ||
+				(*dest == '>') ||
+				(*dest == '|')
+			) *dest = '_';
+#endif
 		}
 
 		if (*src == '\0')
