Description: Fix offset data type in examples.
Author: Martin Pitt <martin.pitt@ubuntu.com>
Bug: http://code.google.com/p/libarchive/issues/detail?id=282

Index: libarchive/examples/minitar/minitar.c
===================================================================
--- libarchive.orig/examples/minitar/minitar.c	2012-11-20 14:59:08.000000000 +0100
+++ libarchive/examples/minitar/minitar.c	2012-11-20 14:59:32.225024976 +0100
@@ -400,7 +400,7 @@
 	int r;
 	const void *buff;
 	size_t size;
-	off_t offset;
+	uint64_t offset;
 
 	for (;;) {
 		r = archive_read_data_block(ar, &buff, &size, &offset);
Index: libarchive/examples/untar.c
===================================================================
--- libarchive.orig/examples/untar.c	2012-11-20 14:59:08.000000000 +0100
+++ libarchive/examples/untar.c	2012-11-20 14:59:50.461025670 +0100
@@ -200,7 +200,7 @@
 	int r;
 	const void *buff;
 	size_t size;
-#if ARCHIVE_VERSION >= 3000000
+#if ARCHIVE_VERSION_NUMBER >= 3000000
 	int64_t offset;
 #else
 	off_t offset;
