--- -	2009-04-16 20:36:26.943727997 +0200
+++ test_option_t.c	2009-04-16 20:35:56.851486409 +0200
@@ -67,7 +67,7 @@
 	/* assertEqualFile("tv.out", "test_option_tv.stdout"); */
 
 	/* List reference archive with numeric IDs, verify TOC is correct. */
-	r = systemf("%s -itnv < test_option_t.cpio >itnv.out 2>itnv.err",
+	r = systemf("/usr/bin/env TZ=UTC %s -itnv < test_option_t.cpio >itnv.out 2>itnv.err",
 		    testprog);
 	assertEqualInt(r, 0);
 	assertTextFileContents("1 block\n", "itnv.err");
