Index: archive_write_set_format_7zip.c
===================================================================
--- archive_write_set_format_7zip.c	(revision 4217)
+++ archive_write_set_format_7zip.c	(working copy)
@@ -405,7 +405,7 @@
 		    value[1] != '\0') {
 			archive_set_error(&(a->archive),
 			    ARCHIVE_ERRNO_MISC,
-			    "Illeagal value `%s'",
+			    "Illegal value `%s'",
 			    value);
 			return (ARCHIVE_FAILED);
 		}
@@ -413,6 +413,10 @@
 		return (ARCHIVE_OK);
 	}
 
+	archive_set_error(&(a->archive),
+		ARCHIVE_ERRNO_MISC,
+		"Illegal option `%s'",
+		value);
 	return (ARCHIVE_FAILED);
 }
 
