Index: libarchive/archive_write_set_format_zip.c
===================================================================
--- libarchive/archive_write_set_format_zip.c	(revision 2954)
+++ libarchive/archive_write_set_format_zip.c	(working copy)
@@ -502,6 +502,9 @@
 	int entries;
 	int ret;
 
+	if (a->compressor.write == NULL)
+		return (ARCHIVE_OK);
+
 	zip = a->format_data;
 	l = zip->central_directory;
 
