Index: libarchive/archive_write_disk_windows.c
===================================================================
--- libarchive/archive_write_disk_windows.c	(revision 3688)
+++ libarchive/archive_write_disk_windows.c	(working copy)
@@ -1883,7 +1883,11 @@
 		return (ARCHIVE_FAILED);
 	}
 
+	/* Accept leading drive identifier. */
+	if (src[0] >= 'A' && src[0] <= 'Z' && src[1] == ':')
+		src += 2;
+
-	for (p = a->name; *p != L'\0'; p++) {
+	for (p = src; *p != L'\0'; p++) {
 		if (*p == L'/')
 			*p = L'\\';
 		/* Rewrite the path name if its character is a unusable. */
