diff --git a/libarchive/test/test_sparse_basic.c b/libarchive/test/test_sparse_basic.c
index 52d76c6..6910c76 100644
--- a/libarchive/test/test_sparse_basic.c
+++ b/libarchive/test/test_sparse_basic.c
@@ -204,11 +204,9 @@ is_sparse_supported(const char *path)
 	fm->fm_extent_count = (sizeof(buff) - sizeof(*fm))/
 		sizeof(struct fiemap_extent);
 	r = ioctl(fd, FS_IOC_FIEMAP, fm);
-	if (r < 0 && (errno == ENOTTY || errno == EOPNOTSUPP))
-		return (0);/* Not supported. */
 	close(fd);
 	unlink(testfile);
-	return (1);
+	return (r >= 0);
 }
 
 #else
