Index: cpio/test/test_owner_parse.c
===================================================================
--- cpio/test/test_owner_parse.c	(revision 2269)
+++ cpio/test/test_owner_parse.c	(working copy)
@@ -31,7 +31,7 @@
 #if !defined(_WIN32)
 #define ROOT "root"
 static int root_uids[] = { 0 };
-static int root_gids[] = { 0 };
+static int root_gids[] = { 0, 1 };
 #elif defined(__CYGWIN__)
 /* On cygwin, the Administrator user most likely exists (unless
  * it has been renamed or is in a non-English localization), but
