Index: archive_read_support_format_iso9660.c
===================================================================
--- archive_read_support_format_iso9660.c	(revision 4013)
+++ archive_read_support_format_iso9660.c	(working copy)
@@ -1962,7 +1962,7 @@
 		    (parent->re || parent->re_descendant))
 			file->re_descendant = 1;
 		if (file->cl_offset) {
-			struct file_info *p;
+			struct file_info *r;
 
 			if (parent == NULL || parent->parent == NULL) {
 				archive_set_error(&a->archive,
@@ -1990,8 +1990,8 @@
 			 * Sanity check: cl_offset does not point at its
 			 * the parents or itself.
 			 */
-			for (p = parent; p; p = p->parent) {
-				if (p->offset == file->cl_offset) {
+			for (r = parent; r; r = r->parent) {
+				if (r->offset == file->cl_offset) {
 					archive_set_error(&a->archive,
 					    ARCHIVE_ERRNO_MISC,
 					    "Invalid Rockridge CL");
